### PR TITLE
WIP: fix: variation select

### PR DESCRIPTION
**Pull request by Sebastian-Haehnlein**
## PR Type

 [x] Bugfix  
 [x] Feature  

## What Is the New Behavior?

Bugfix in case of one variation attribute all options were "not available" any time. Additional new behaviour in case of one variation attribute is hiding "not available" options. (ISREST-756)  

"not available" selections should not be possible. (ISREST-725 & ISREST-1019)  
Now it is possible to find a possible variation based on the changed attribute.

In case of just one variation attribute and only one option now it reloads the component/page (second commit & cornercase)

## Does this PR Introduce a Breaking Change?

 [ ] Yes  
 [x] No

### CODE REVIEW

File: src/app/shared/components/product/product-variation-select/product-variation-select.component.ts [hunk 1]
```diff
@@ -36,10 +38,22 @@ export class ProductVariationSelectComponent implements OnChanges, OnDestroy {
...
+      });
     }
   }
 
+  getChangedAttribute(selection) {
+    const diff = reduce(
+      selection,
+      (result, value, key) => (isEqual(value, this.initialSelection[key]) ? result : result.concat(key)),
+      []
+    );
```
Comment (dhhyi) [inline src/app/shared/components/product/product-variation-select/product-variation-select.component.ts:53]: What is the advantage of using `reduce` from `lodash` instead of the built-in `selection.reduce`?

File: src/app/shared/components/product/product-item/product-item.component.ts [hunk 2]
```diff
@@ -107,6 +108,9 @@ export class ProductItemComponent implements OnInit, OnChanges, OnDestroy {
     this.loading$ = this.shoppingFacade.productNotReady$(this.sku$, ProductItemComponent.REQUIRED_COMPLETENESS_LEVEL);
 
     this.productVariationOptions$ = this.shoppingFacade.productVariationOptions$(this.sku$);
+    this.productVariationOptions$
+      .pipe(takeUntil(this.destroy$))
+      .subscribe(x => (this.productVariationOptions = x ? [...x] : x)); // trigger change detection
```
Comment (dhhyi) [inline src/app/shared/components/product/product-item/product-item.component.ts:113]: For my understanding: Why is this necessary? What does this fix?

File: src/app/shared/components/line-item/line-item-edit-dialog/line-item-edit-dialog.component.ts [hunk 3]
```diff
@@ -89,6 +90,7 @@ export class LineItemEditDialogComponent implements OnInit, OnDestroy, OnChanges
     this.product$ = this.shoppingFacade.product$(this.sku$, ProductCompletenessLevel.List);
 
     this.variationOptions$ = this.shoppingFacade.productVariationOptions$(this.sku$);
+    this.variationOptions$.pipe(takeUntil(this.destroy$)).subscribe(x => (this.variationOptions = x ? [...x] : x)); // trigger change detection
```
Comment (dhhyi) [inline src/app/shared/components/line-item/line-item-edit-dialog/line-item-edit-dialog.component.ts:93]: For my understanding: Why is this necessary? What does this fix?

File: src/app/pages/product/product-page.component.ts [hunk 4]
```diff
@@ -142,7 +149,9 @@ export class ProductPageComponent implements OnInit, OnDestroy {
       )
       .subscribe(([product, category]) => {
         this.ngZone.run(() => {
-          this.router.navigateByUrl(generateProductUrl(product, category), { replaceUrl });
+          this.router.navigateByUrl('/', { skipLocationChange: true }).then(() => {
+            this.router.navigateByUrl(generateProductUrl(product, category), { replaceUrl });
+          });
```
Comment (dhhyi) [inline src/app/pages/product/product-page.component.ts:154]: For my understanding: Why is this necessary? What does this fix?

File: src/app/core/models/product-variation/product-variation.helper.ts [hunk 5]
```diff
@@ -109,37 +102,52 @@ export class ProductVariationHelper {
   // TODO: Refactor this to a more functional style
   static findPossibleVariationForSelection(
```
Comment (dhhyi) [inline src/app/core/models/product-variation/product-variation.helper.ts:103]: please add testing

File: src/app/pages/product/product-page.component.ts (hunk 4 above)
Comment (Sebastian-Haehnlein) [reply src/app/pages/product/product-page.component.ts:154]: in case of redirect to itself (found by findPossibleVariationForSelection), the page lifecycle does not work. (corner case)

File: src/app/shared/components/line-item/line-item-edit-dialog/line-item-edit-dialog.component.ts (hunk 3 above)
Comment (Sebastian-Haehnlein) [reply src/app/shared/components/line-item/line-item-edit-dialog/line-item-edit-dialog.component.ts:93]: in case of redirect to itself (found by findPossibleVariationForSelection), the component lifecycle does not work. (reset selection), so i have to be sure new inputs with same values triggers ngchanges. 

(i dont know, is there a copy operator or somethime in rxjs)

File: src/app/pages/product/product-page.component.ts (hunk 4 above)
Comment (dhhyi) [reply src/app/pages/product/product-page.component.ts:154]: Could you try activating this? https://angular.io/api/router/ExtraOptions#onSameUrlNavigation

Comment (Sebastian-Haehnlein) [reply src/app/pages/product/product-page.component.ts:154]: i tryed onSameUrlNavigation, but it doesnt work
https://drive.google.com/file/d/1TKxCFneonFgCR12o7SAJeZ6PxBEVKxhW/view
